### PR TITLE
CC-1772: Feature/warn when state is in disarray

### PR DESCRIPTION
**Pull request by knite51**
Expanded on drafted PR: https://github.com/companieshouse/chs-dev/pull/112

[Add post and pre hook state checks](https://github.com/companieshouse/chs-dev/commit/150a18eb442caa66d749098bd4a5045a030543c2) 

[Add functionality to check the state of services and modules](https://github.com/companieshouse/chs-dev/commit/61b7f5a298b4daf03f5fe34e51e18020bcc0908a) 

[add general helper array related functions](https://github.com/companieshouse/chs-dev/commit/ce41c4b7e063369b54d1adc4e1ade061a92fb646)

[Adhere to Dry principles in the Service loader class](https://github.com/companieshouse/chs-dev/commit/a9761a1c0afbacb63c82330fb498a0be7f9a8d22) 

[Add unit test for the commands ](https://github.com/companieshouse/chs-dev/commit/062240fd6e1cebf2d8f88cf2f2b1079e00f5c171) 
[add unit test for the hook, helper func and serviceloader class](https://github.com/companieshouse/chs-dev/commit/2da04f41b126cfee4154545889e0200aeb98c0fd)

Edge Cases are
- Prompt when enabling service or module but services in devmode or exclusion list are not enabled.
- Prompt and stop command when enabling dev mode service but the service in question was not activated before the development command was initiated.
- Prompt and stop command when adding service to exclusion list  but the service in question was not activated before the exlcusion command was initiated.

Each of the prompt has the option of removing the troublesome service.

### CODE REVIEW

File: src/commands/AbstractStateModificationCommand.ts [hunk 1]
```diff
@@ -78,7 +92,63 @@ export default abstract class AbstractStateModificationCommand extends Command {
...
+    protected handleActivatedServicesByName (): string[] {
+        const state = this.stateManager.snapshot;
+        const enabledServiceNames = new ServiceLoader(this.inventory);
+        return enabledServiceNames.loadServicesNames(state);
+    }
+
+    private handleHookResponse (result): string[] {
+        let hookWarnings: string[] = [];
+        for (const success of result.successes) {
+            hookWarnings = success.result;
```
Comment (NeilShadrach) [inline src/commands/AbstractStateModificationCommand.ts:147]: Isn't this overwriting hookWarnings each time rather than appending to it?

File: test/hooks/check-service-or-module-state.spec.ts [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    projectName: "chs-dev-proj",
+    projectPath: "/home/test-user/projects/chs-dev-proj",
+    env: {}
+};
+
+jest.mock("../../src/state/state-manager");
+jest.mock("../../src/state/inventory");
+jest.mock("../../src/helpers/config-loader");
+jest.mock("../../src/helpers/user-input");
+// jest.mock("../../src/helpers/index");
```
Comment (NeilShadrach) [inline test/hooks/check-service-or-module-state.spec.ts:49]: Remove commented out code

File: test/hooks/check-service-or-module-state.spec.ts [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+        StateManager.mockReturnValue(stateManagerMock);
+
+        // @ts-expect-error
+        Inventory.mockReturnValue(inventoryMock);
+
+        // @ts-expect-error
+        configLoaderMock.mockReturnValue(testConfig);
+
+        (input as jest.Mock).mockImplementation(() => "no");
+        // (matchExistInArrays as jest.Mock).mockReturnValue(false);
```
Comment (NeilShadrach) [inline test/hooks/check-service-or-module-state.spec.ts:72]: Remove commented out code

File: src/commands/AbstractStateModificationCommand.ts (hunk 1 above)
Comment (knite51) [reply src/commands/AbstractStateModificationCommand.ts:147]: Yup. Thats the idea.

File: test/hooks/check-service-or-module-state.spec.ts (hunk 2 above)
Comment (knite51) [reply test/hooks/check-service-or-module-state.spec.ts:49]: WIP. My bad. thanks

File: src/commands/AbstractStateModificationCommand.ts (hunk 1 above)
Comment (NeilShadrach) [reply src/commands/AbstractStateModificationCommand.ts:147]: Fair enough.  Might be better to explicitly set it to the last result if present rather than loop in that case

File: src/helpers/index.ts [hunk 4]
```diff
@@ -17,3 +17,25 @@ export const isObjectFormat = (T: unknown): boolean => {
...
+/**
+ * Finds and returns elements from the first array that do not exist in the second array.
+ *
+ * @param {unknown[]} array1 - The source array to check.
+ * @param {unknown[]} array2 - The reference array to compare against.
+ * @returns {any[]} - An array containing elements from `array1` that are not present in `array2`.
+ */
+export const findUniqueItems = (array1: unknown[], array2: unknown[]): any[] => {
+    return array1.filter(item => !array2.includes(item));
+};
```
[the remark this answers is not in the document]
Comment (knite51) [reply src/helpers/index.ts:41]: Yes, totally skipped my mind. Thanks

File: src/hooks/check-service-or-module-state.ts [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+        handler = handleDevelopmentExclusionsTopic(activatedServices as string[], commandArgv as string[]);
+        break;
+
+    }
+
+    if (typeof handler !== "undefined") {
+        handler(command as string, config as Config, context);
+    }
+
+    return warnings;
```
[the remark this answers is not in the document]
Comment (knite51) [reply src/hooks/check-service-or-module-state.ts:48]: Big thanks for this. You know when you know something is off about your code but cant think it out or figure a refactor. Such obvious solution. 🤦

File: src/commands/AbstractStateModificationCommand.ts [hunk 6]
```diff
@@ -141,8 +142,8 @@ export default abstract class AbstractStateModificationCommand extends Command {
         return enabledServiceNames.loadServicesNames(state);
     }
 
-    private handleHookResponse (result): string[] {
-        return result.successes.flatMap(success => success.result);
+    private handleHookResponse (result): string | undefined {
```
Comment (JoelClemence) [inline src/commands/AbstractStateModificationCommand.ts:145]: is there a type for `result` or is it ok being implicit any?

**Review by JoelClemence — APPROVED** [review]
Very minor comment regarding typing a parameter feel free to take or leave it

File: src/commands/AbstractStateModificationCommand.ts (hunk 6 above)
Comment (knite51) [reply src/commands/AbstractStateModificationCommand.ts:145]: There should be. I have checked the documentation I couldn't find the type definitions for hook responses. I had to console log result to see the object properties.

Comment (JoelClemence) [reply src/commands/AbstractStateModificationCommand.ts:139]: I meant for this method

given:
```ts
private handleHookResponse (result): string | undefined
```

the `result` parameter has implicit `any` type - I was wondering whether you can give this parameter a type?

**Review by JoelClemence — APPROVED** [review]

**Review by JoelClemence — APPROVED** [review]
